### PR TITLE
[Storage] Rename File

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs [hunk 1]
```diff
@@ -788,6 +792,20 @@ public enum ShareFileRangeWriteType
         Update = 0,
         Clear = 1,
     }
+    public partial class ShareFileRenameOptions
+    {
+        public ShareFileRenameOptions() { }
+        public string ContentType { get { throw null; } set { } }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs:798]: `ShareFileRenameOptions` is used for both File and Directory Rename.  `ContentType` is not settable for a directory, so maybe we should throw an ArguementException in `ShareDirectoryClient.Rename()` if the user attempts to set it?

File: sdk/storage/Azure.Storage.Files.Shares/src/autorest.md [hunk 2]
```diff
@@ -4,7 +4,7 @@ Run `dotnet build /t:GenerateCode` to generate code.
 
 ``` yaml
 input-file:
-    - https://raw.githubusercontent.com/Azure/azure-rest-api-specs/60723d13309c8f8060d020a7f3dd9d6e380f0bbd/specification/storage/data-plane/Microsoft.FileStorage/preview/2021-02-12/file.json
+    - C:\Users\seanmcc\git\azure-rest-api-specs\specification\storage\data-plane\Microsoft.FileStorage\preview\2021-04-10\file.json
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/autorest.md:7]: TODO fix this before merging

File: sdk/storage/Azure.Storage.Files.Shares/tests/DirectoryClientTests.cs [hunk 3]
```diff
@@ -1228,6 +1229,235 @@ await foreach (ShareFileItem shareFileItem in directoryClient.GetFilesAndDirecto
...
+
+            // Assert
+            await destDirectory.GetPropertiesAsync();
+        }
+
+        [RecordedTest]
+        [ServiceVersion(Min = ShareClientOptions.ServiceVersion.V2021_04_10)]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task RenameAsync_ReplaceIfExists(bool replaceIfExists)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/DirectoryClientTests.cs:1276]: Note that apparently it is legal to overwrite a file with a directory, but not a directory with another directory.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileRenameOptions.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Metadata = System.Collections.Generic.IDictionary<string, string>;
+
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Optional parameters for renaming a File or Directory.
+    /// </summary>
+    public class ShareFileRenameOptions
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileRenameOptions.cs:14]: If these options are going to be used for both ShareFile and ShareDirectory, should the name be more like SharePathRenameOptions? I don't have a strong opinion on the name.

File: sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs (hunk 1 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs:798]: If `ContentType` is not settable for Directories, shouldn't there be two different option bags then? `ShareFileRenameOptions` and `ShareDirectoryRenameOptions` that way the user knows what parameters can be set?

Some customers see the SDK as the definition of what the service accepts (not that I am approving of why they think like this).

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs:798]: @amnguye, maybe we could just pull `ContentType` from this type?  I guess Rename Directory and Rename File could also further diverge, so maybe we should make 2 different types.

File: sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs [hunk 5]
```diff
@@ -788,6 +792,20 @@ public enum ShareFileRangeWriteType
...
+        public Azure.Storage.Files.Shares.Models.ShareFileRequestConditions DestinationRequestConditions { get { throw null; } set { } }
+        public Azure.Storage.Files.Shares.Models.NtfsFileAttributes? FileAttributes { get { throw null; } set { } }
+        public System.DateTimeOffset? FileCreatedOn { get { throw null; } set { } }
+        public System.DateTimeOffset? FileLastWrittenOn { get { throw null; } set { } }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/api/Azure.Storage.Files.Shares.netstandard2.0.cs:802]: The spec also mentions `x-ms-file-change-time`, which we are missing in this class.

File: sdk/storage/Azure.Storage.Files.Shares/src/autorest.md (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/autorest.md:7]: Still need to fix this.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileRenameOptions.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileRenameOptions.cs:14]: Let's see what the arch board says.

**Review by amnguye — APPROVED** [review]

**Pull request merged**